File: nihongo-it-backend/src/main/kotlin/com/example/nihongoit/service/AdminService.kt
package com.example.nihongoit.service

import com.example.nihongoit.dto.admin.UserListResponse
import com.example.nihongoit.dto.user.UserCreateRequest
import com.example.nihongoit.dto.user.UserDto
import com.example.nihongoit.dto.user.UserUpdateRequest
import com.example.nihongoit.entity.UserEntity
import com.example.nihongoit.exception.BusinessException
import com.example.nihongoit.repository.RoleRepository
import com.example.nihongoit.repository.UserRepository
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.util.*

@Service
class AdminService(
    private val userRepository: UserRepository,
    private val roleRepository: RoleRepository,
    private val passwordEncoder: PasswordEncoder
) {
    private val logger = LoggerFactory.getLogger(AdminService::class.java)
    
    /**
     * Get all users with pagination and optional search
     */
    fun getAllUsers(pageable: Pageable, search: String?): UserListResponse {
        val userPage: Page<UserEntity> = if (search.isNullOrBlank()) {
            userRepository.findAll(pageable)
        } else {
            // Use a custom repository method to search by email or fullName
            // This is a basic implementation, you may need to create a custom query
            userRepository.findByEmailContainingIgnoreCaseOrFullNameContainingIgnoreCase(
                search, search, pageable
            )
        }
        
        val users = userPage.content.map { it.toUserDto() }
        
        return UserListResponse(
            users = users,
            totalItems = userPage.totalElements,
            totalPages = userPage.totalPages,
            currentPage = pageable.pageNumber
        )
    }
    
    /**
     * Get user by ID
     */
    fun getUserById(userId: UUID): UserDto {
        val user = findUserById(userId)
        return user.toUserDto()
    }
    
    /**
     * Create a new user
     */
    @Transactional
    fun createUser(request: UserCreateRequest): UserDto {
        // Check if email already exists
        if (userRepository.existsByEmail(request.email)) {
            throw BusinessException("Email ${request.email} is already registered")
        }
        
        // Get role
        val role = roleRepository.findByRoleId(request.roleId)
            ?: throw BusinessException("Invalid role ID: ${request.roleId}")
        
        // Create user entity
        val newUser = UserEntity(
            email = request.email,
            password = passwordEncoder.encode(request.password),
            fullName = request.fullName,
            profilePicture = request.profilePicture,
            currentLevel = request.currentLevel,
            jlptGoal = request.jlptGoal,
            isActive = true,
            isEmailVerified = true, // Admin-created accounts are pre-verified
            lastLogin = null,
            role = role,
            createdAt = LocalDateTime.now(),
            updatedAt = LocalDateTime.now()
        )
        
        val savedUser = userRepository.save(newUser)
        logger.info("User created by admin: ${savedUser.email}")
        
        return savedUser.toUserDto()
    }
    
    /**
     * Update an existing user
     */
    @Transactional
    fun updateUser(userId: UUID, request: UserUpdateRequest): UserDto {
        val existingUser = findUserById(userId)
        
        // Check if email is being changed and already exists
        if (request.email != null && 
            request.email != existingUser.email && 
            userRepository.existsByEmail(request.email)) {
            throw BusinessException("Email ${request.email} is already registered")
        }
        
        // Prepare role if it's being updated
        val role = if (request.roleId != null) {
            roleRepository.findByRoleId(request.roleId)
                ?: throw BusinessException("Invalid role ID: ${request.roleId}")
        } else {
            existingUser.role
        }
        
        // Parse reminder time if provided
        val reminderTime = if (request.reminderTime != null) {
            try {
                LocalTime.parse(request.reminderTime, DateTimeFormatter.ofPattern("HH:mm"))
            } catch (e: Exception) {
                throw BusinessException("Invalid reminder time format. Use HH:mm")
            }
        } else {
            existingUser.reminderTime
        }
        
        // Update user with non-null values from request
        val updatedUser = existingUser.copy(
            email = request.email ?: existingUser.email,
            password = if (request.password != null) 
                passwordEncoder.encode(request.password) else existingUser.password,
            fullName = request.fullName ?: existingUser.fullName,
            profilePicture = request.profilePicture ?: existingUser.profilePicture,
            currentLevel = request.currentLevel ?: existingUser.currentLevel,
            jlptGoal = request.jlptGoal ?: existingUser.jlptGoal,
            isActive = request.isActive ?: existingUser.isActive,
            isEmailVerified = request.isEmailVerified ?: existingUser.isEmailVerified,
            reminderEnabled = request.reminderEnabled ?: existingUser.reminderEnabled,
            reminderTime = reminderTime,
            notificationPreferences = request.notificationPreferences 
                ?: existingUser.notificationPreferences,
            minCardThreshold = request.minCardThreshold ?: existingUser.minCardThreshold,
            role = role,
            updatedAt = LocalDateTime.now()
        )
        
        val savedUser = userRepository.save(updatedUser)
        logger.info("User updated by admin: ${savedUser.email}")
        
        return savedUser.toUserDto()
    }
    
    /**
     * Deactivate a user (soft delete)
     */
    @Transactional
    fun deactivateUser(userId: UUID) {
        val user = findUserById(userId)
        
        // Don't allow deactivation of the last admin user
        if (user.role.roleId == 1) {
            val adminCount = userRepository.countByRoleRoleIdAndIsActive(1, true)
            if (adminCount <= 1) {
                throw BusinessException("Cannot deactivate the last active admin user")
            }
        }
        
        val deactivatedUser = user.copy(
            isActive = false,
            updatedAt = LocalDateTime.now()
        )
        
        userRepository.save(deactivatedUser)
        logger.info("User deactivated by admin: ${user.email}")
    }
    
    /**
     * Activate a user
     */
    @Transactional
    fun activateUser(userId: UUID) {
        val user = findUserById(userId)
        
        val activatedUser = user.copy(
            isActive = true,
            updatedAt = LocalDateTime.now()
        )
        
        userRepository.save(activatedUser)
        logger.info("User activated by admin: ${user.email}")
    }
    
    /**
     * Change a user's role
     */
    @Transactional
    fun changeUserRole(userId: UUID, roleId: Int) {
        val user = findUserById(userId)
        
        // Get the new role
        val newRole = roleRepository.findByRoleId(roleId)
            ?: throw BusinessException("Invalid role ID: $roleId")
        
        // Don't allow changing the last admin to a regular user
        if (user.role.roleId == 1 && roleId != 1) {
            val adminCount = userRepository.countByRoleRoleIdAndIsActive(1, true)
            if (adminCount <= 1) {
                throw BusinessException("Cannot change role of the last active admin user")
            }
        }
        
        val updatedUser = user.copy(
            role = newRole,
            updatedAt = LocalDateTime.now()
        )
        
        userRepository.save(updatedUser)
        logger.info("User role changed by admin: ${user.email}, new role: ${newRole.roleName}")
    }
    
    /**
     * Helper method to find a user by ID or throw exception
     */
    private fun findUserById(userId: UUID): UserEntity {
        return userRepository.findById(userId)
            .orElseThrow { BusinessException("User not found with ID: $userId") }
    }
    
    /**
     * Extension function to convert UserEntity to UserDto
     */
    private fun UserEntity.toUserDto(): UserDto {
        return UserDto(
            userId = this.userId ?: throw BusinessException("User ID is null"),
            email = this.email,
            fullName = this.fullName,
            roleId = this.role.roleId,
            profilePicture = this.profilePicture,
            currentLevel = this.currentLevel,
            jlptGoal = this.jlptGoal,
            lastLogin = this.lastLogin,
            isActive = this.isActive
        )
    }
} 

File: frontend/src/services/admin.service.ts
import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8080';

// Create axios instance with authorization header interceptor
const apiClient = axios.create({
  baseURL: API_URL
});

// Add a request interceptor to include the token in all requests
apiClient.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('auth_token');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

export interface UserCreateRequest {
  email: string;
  password: string;
  fullName: string;
  profilePicture?: string;
  currentLevel?: string;
  jlptGoal?: string;
  roleId?: number;
}

export interface UserUpdateRequest {
  email?: string;
  password?: string;
  fullName?: string;
  profilePicture?: string;
  currentLevel?: string;
  jlptGoal?: string;
  isActive?: boolean;
  isEmailVerified?: boolean;
  roleId?: number;
  reminderEnabled?: boolean;
  reminderTime?: string;
  notificationPreferences?: string;
  minCardThreshold?: number;
}

export interface UserInfo {
  userId: string;
  email: string;
  fullName: string;
  roleId: number;
  profilePicture?: string;
  currentLevel?: string;
  jlptGoal?: string;
  lastLogin?: string;
  isActive?: boolean;
  isEmailVerified?: boolean;
  createdAt?: string;
  updatedAt?: string;
  streakCount?: number;
  points?: number;
  reminderEnabled?: boolean;
  reminderTime?: string;
  notificationPreferences?: string;
  minCardThreshold?: number;
  flashcardCount?: number;
  newCards?: number;
  learningCards?: number;
  masteredCards?: number;
}

export interface UserListResponse {
  users: UserInfo[];
  totalItems: number;
  totalPages: number;
  currentPage: number;
}

export interface DashboardStats {
  userCount: number;
  vocabularyCount: number;
  categoryCount: number;
  topicCount: number;
  newUsers: number;
  activeUsers: number;
  searchesToday: number;
  flashcardsCreatedToday: number;
  flashcardsStudiedToday: number;
  recentActivities: {
    user: string;
    action: string;
    timestamp: string;
  }[];
}

class AdminService {
  /**
   * Get dashboard statistics
   */
  async getDashboardStats(): Promise<DashboardStats> {
    try {
      const response = await apiClient.get('/api/admin/dashboard/stats');
      return response.data.data;
    } catch (error) {
      console.error('Error fetching dashboard stats:', error);
      throw error;
    }
  }

  /**
   * Get all users with pagination and optional search
   */
  async getUsers(page = 0, size = 10, search?: string, sortBy = 'email', sortDir = 'asc'): Promise<UserListResponse> {
    try {
      const params = { page, size, sortBy, sortDir };
      if (search) {
        Object.assign(params, { search });
      }
      const response = await apiClient.get('/api/admin/users', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching users:', error);
      throw error;
    }
  }

  /**
   * Get user by ID
   */
  async getUserById(userId: string): Promise<UserInfo> {
    try {
      const response = await apiClient.get(`/api/admin/users/${userId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching user:', error);
      throw error;
    }
  }

  /**
   * Create a new user
   */
  async createUser(userData: UserCreateRequest): Promise<UserInfo> {
    try {
      const response = await apiClient.post('/api/admin/users', userData);
      return response.data;
    } catch (error) {
      console.error('Error creating user:', error);
      throw error;
    }
  }

  /**
   * Update an existing user
   */
  async updateUser(userId: string, userData: UserUpdateRequest): Promise<UserInfo> {
    try {
      const response = await apiClient.put(`/api/admin/users/${userId}`, userData);
      return response.data;
    } catch (error) {
      console.error('Error updating user:', error);
      throw error;
    }
  }

  /**
   * Deactivate a user (soft delete)
   */
  async deactivateUser(userId: string): Promise<{ status: string; message: string }> {
    try {
      const response = await apiClient.delete(`/api/admin/users/${userId}`);
      return response.data;
    } catch (error) {
      console.error('Error deactivating user:', error);
      throw error;
    }
  }

  /**
   * Activate a user
   */
  async activateUser(userId: string): Promise<{ status: string; message: string }> {
    try {
      const response = await apiClient.put(`/api/admin/users/${userId}/activate`);
      return response.data;
    } catch (error) {
      console.error('Error activating user:', error);
      throw error;
    }
  }

  /**
   * Change a user's role
   */
  async changeUserRole(userId: string, roleId: number): Promise<{ status: string; message: string }> {
    try {
      const response = await apiClient.put(`/api/admin/users/${userId}/change-role`, null, {
        params: { roleId }
      });
      return response.data;
    } catch (error) {
      console.error('Error changing user role:', error);
      throw error;
    }
  }
}

export default new AdminService();
